README.md
# ui-audit

Copyright (C) 2018 The Open Library Foundation

This software is distributed under the terms of the Apache License, Version 2.0. See the file "[LICENSE](LICENSE)" for more information.

## Introduction

Stripes UI module- Audit app  visually conveys the idea of examining past activity, organized by different points of view: user; resource; circulation etc

## Prerequisites

In order to view and log into the platform being served up, a suitable Okapi backend will need to be running. The [Folio testing-backend](https://app.vagrantup.com/folio/boxes/testing-backend) Vagrant box should work if your app does not yet have its own backend module.

## Run your new app

Run the following from the ui-audit directory to serve your new app using a development server:
```
yarn start

or

stripes serve
```

Note: When serving up a newly created app that does not have its own backend permissions established, pass the `--hasAllPerms` option to display the app in the UI navigation. For example:
```
yarn start --hasAllPerms

or

stripes serve --hasAllPerms
```

To specify your own tenant ID or to use an Okapi instance other than http://localhost:9130, pass the `--okapi` and `--tenant` options.
```
yarn start --okapi http://my-okapi.example.com:9130 --tenant my-tenant-id

or

stripes serve --okapi http://my-okapi.example.com:9130 --tenant my-tenant-id
```

## Run the tests

Run the included UI test `demo` with the following command without electron:
```
yarn test

or

stripes test --run demo
```
Run the included UI test `demo` with the following command:
```
yarn test --show

or

stripes test --run demo --show
```

CHANGELOG.md
# Change history for ui-audit

## 1.0.1 (10/4/2018)

* Upgrade stripes v1

## 1.0.0 (10/4/2018)

* Very simple audit data capture module


test/ui-testing/demo.js
/* global Nightmare, describe, it, before, after */
const chai = require('chai');

const expect = chai.expect;
module.exports.test = (uiTestCtx) => {
  describe('Module test: ui-audit:', function startTest() {
    const { config, helpers: { login, logout } } = uiTestCtx;
    const nightmare = new Nightmare(config.nightmare);
    this.timeout(Number(config.test_timeout));

    describe('Login > navigate to app > verify message > logout', () => {
      const testRunYear = new Date().getFullYear();
      before((done) => {
        login(nightmare, config, done);
      });
      after((done) => {
        logout(nightmare, config, done);
      });
      it('should open app and search for login user ', (done) => {
        nightmare
          .wait('#clickable-audit-module')
          .click('#clickable-audit-module')
          .wait('#input-audit-search')
          .click('#input-audit-search')
          .insert('#input-audit-search', config.username)
          .wait('#clickable-list-column-timestamp')
          .click('#paneHeaderpane-results-subtitle')
          .evaluate(() => { return document.querySelector('#list-audit') ? 1 : 0; })
          .then((result) => {
            if (result) {
              nightmare
                .wait('#list-audit div[class*=scrollable] div div div')
                .click('#list-audit div[class*=scrollable] div div div a')
                .click('#audit-module-display div section:nth-child(4) :last-child')
                .then(done)
                .catch(done);
            }
          })
          .catch(error => {
            throw new Error(`Error search on target id with ${config.username}`, error);
          }, done);
      });
      it('should change filter index between all the HTTP methods ', (done) => {
        nightmare
          .wait('#clickable-audit-module')
          .click('#clickable-audit-module')
          .wait('#clickable-filter-method-POST')
          .click('#clickable-filter-method-POST')
          .wait('#clickable-list-column-timestamp')
          .click('#clickable-filter-method-POST')
          .wait('#clickable-list-column-timestamp')
          .click('#clickable-filter-method-PUT')
          .wait('#clickable-list-column-timestamp')
          .click('#clickable-filter-method-PUT')
          .wait('#clickable-list-column-timestamp')
          .click('#clickable-filter-method-DELETE')
          .wait('#clickable-list-column-timestamp')
          .click('#clickable-filter-method-DELETE')
          .wait('#clickable-list-column-timestamp')
          .click('#clickable-filter-method-PATCH')
          .wait('div[class*="noResultsMessage"]')
          .evaluate(() => document.querySelector('div[class*="noResultsMessage"]').innerText.trim())
          .then((result) => {
            expect(result).to.equal(`No results found for "${config.username}". Please check your spelling and filters.`);
            done();
          })
          .catch(done);
      });
      it('should change search index to timestamp ', (done) => {
        nightmare
          .wait('#clickable-audit-module')
          .click('#clickable-audit-module')
          .click('#clickable-reset-all')
          .wait('#input-audit-search-qindex')
          .click('#input-audit-search-qindex')
          .type('#input-audit-search-qindex', 'time')
          .insert('#input-audit-search', testRunYear)
          .wait('#clickable-list-column-timestamp')
          .then(done)
          .catch(done);
      });
      it('should change filter index for HTTP statuses ', (done) => {
        nightmare
          .check('input[id$="201"]')
          .wait('#clickable-list-column-timestamp')
          .uncheck('input[id$="201"]')
          .check('input[id$="204"]')
          .wait('#clickable-list-column-timestamp')
          .uncheck('input[id$="204"]')
          .check('input[id$="403"]')
          .wait('#clickable-list-column-timestamp')
          .uncheck('input[id$="403"]')
          .check('input[id$="400"]')
          .wait('#clickable-list-column-timestamp')
          .uncheck('input[id$="400"]')
          .check('input[id$="500"]')
          .wait('#clickable-list-column-timestamp')
          .uncheck('input[id$="500"]')
          .then(done)
          .catch(done);
      });
    });
  });
};


test/ui-testing/test.js
const auditSearch = require('./auditSearch.js');

module.exports.test = function test(uiTestCtx) {
  auditSearch.test(uiTestCtx);
};


test/ui-testing/auditSearch.js
/* global Nightmare, describe, it, before, after */
const chai = require('chai');

const expect = chai.expect;
module.exports.test = (uiTestCtx) => {
  describe('Module test: ui-audit:', function startTest() {
    const { config, helpers: { login, logout } } = uiTestCtx;
    const nightmare = new Nightmare(config.nightmare);
    this.timeout(Number(config.test_timeout));

    describe('Login > navigate to app > verify message > logout', () => {
      const testRunYear = new Date().getFullYear();
      before((done) => {
        login(nightmare, config, done);
      });
      after((done) => {
        logout(nightmare, config, done);
      });
      it('should open app and search for login user ', (done) => {
        nightmare
          .wait('#clickable-audit-module')
          .click('#clickable-audit-module')
          .wait('#input-audit-search')
          .click('#input-audit-search')
          .insert('#input-audit-search', config.username)
          .wait('#clickable-list-column-timestamp')
          .click('#paneHeaderpane-results-subtitle')
          .evaluate(() => { return document.querySelector('#list-audit') ? 1 : 0; })
          .then((result) => {
            if (result) {
              nightmare
                .wait('#list-audit div[class*=scrollable] div div div')
                .click('#list-audit div[class*=scrollable] a')
                .wait('#audit-module-display div div+section+section+section span[class*=hljs-string]')
                .evaluate(() => {
                  const UUIDregexp = /[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/gi;
                  const auditID = document.querySelector('#list-audit div[class*=scrollable] div div div a').getAttribute('href').match(UUIDregexp);
                  const detailViewID = document.querySelector('#audit-module-display div div+section+section+section').innerText.trim().match(UUIDregexp);
                  return { auditID, detailViewID };
                })
                .then((ids) => {
                  expect(ids.auditID[0]).to.equal(ids.detailViewID[0]);
                  nightmare
                    .click('#audit-module-display div section:nth-child(4) :last-child')
                    .then(done)
                    .catch(done);
                })
                .catch(done);
            }
          })
          .catch(error => {
            throw new Error(`Error search on target id with ${config.username}`, error);
          }, done);
      });
      it('should change filter index between all the HTTP methods ', (done) => {
        nightmare
          .wait('#clickable-audit-module')
          .click('#clickable-audit-module')
          .wait('#clickable-filter-method-POST')
          .click('#clickable-filter-method-POST')
          .wait('#clickable-list-column-timestamp')
          .click('#clickable-filter-method-POST')
          .wait('#clickable-list-column-timestamp')
          .click('#clickable-filter-method-PUT')
          .wait('#clickable-list-column-timestamp')
          .click('#clickable-filter-method-PUT')
          .wait('#clickable-list-column-timestamp')
          .click('#clickable-filter-method-DELETE')
          .wait('#clickable-list-column-timestamp')
          .click('#clickable-filter-method-DELETE')
          .wait('#clickable-list-column-timestamp')
          .click('#clickable-filter-method-PATCH')
          .wait('div[class*="noResultsMessage"]')
          .evaluate(() => document.querySelector('div[class*="noResultsMessage"]').innerText.trim())
          .then((result) => {
            expect(result).to.equal(`No results found for "${config.username}". Please check your spelling and filters.`);
            done();
          })
          .catch(done);
      });
      it('should change search index to timestamp ', (done) => {
        nightmare
          .wait('#clickable-audit-module')
          .click('#clickable-audit-module')
          .click('#clickable-reset-all')
          .wait('#input-audit-search-qindex')
          .click('#input-audit-search-qindex')
          .type('#input-audit-search-qindex', 'time')
          .insert('#input-audit-search', testRunYear)
          .wait('#clickable-list-column-timestamp')
          .then(done)
          .catch(done);
      });
      it('should change filter index for HTTP statuses ', (done) => {
        nightmare
          .check('input[id$="201"]')
          .wait('#clickable-list-column-timestamp')
          .uncheck('input[id$="201"]')
          .check('input[id$="204"]')
          .wait('#clickable-list-column-timestamp')
          .uncheck('input[id$="204"]')
          .check('input[id$="403"]')
          .wait('#clickable-list-column-timestamp')
          .uncheck('input[id$="403"]')
          .check('input[id$="400"]')
          .wait('#clickable-list-column-timestamp')
          .uncheck('input[id$="400"]')
          .check('input[id$="500"]')
          .wait('#clickable-list-column-timestamp')
          .uncheck('input[id$="500"]')
          .then(done)
          .catch(done);
      });
    });
  });
};


src/index.js
import React from 'react';
import PropTypes from 'prop-types';
import Switch from 'react-router-dom/Switch';
import Route from 'react-router-dom/Route';
import Search from './routes/search';

const addScript = () => {
  if (!document.getElementById('hightlightJS')) {
    const l = document.createElement('link');
    l.id = 'hightlightJS';
    l.setAttribute('rel', 'stylesheet');
    l.setAttribute('href', 'https://highlightjs.org/static/demo/styles/idea.css');
    document.head.appendChild(l);
  }
};
class Audit extends React.Component {
  static propTypes = {
    match: PropTypes.object.isRequired,
    showSettings: PropTypes.bool,
    location: PropTypes.object.isRequired,
    stripes: PropTypes.shape({
      connect: PropTypes.func.isRequired,
    }).isRequired,
  }

  constructor(props) {
    super(props);
    this.connectedApp = props.stripes.connect(Search);
    addScript();
  }

  render() {
    return (
      <Switch>
        <Route path={`${this.props.match.path}`} render={() => <this.connectedApp {...this.props} />} />
        <Route component={() => <div><p>Uh-oh!</p></div>} />
      </Switch>
    );
  }
}

export default Audit;


src/components/ViewAuditData.js
import React from 'react';
import PropTypes from 'prop-types';
import { Pane, Icon } from '@folio/stripes/components';
import Highlight from 'react-highlight.js';

class ViewAuditData extends React.Component {
  static manifest = Object.freeze({
    query: {},
    selectedInstance: {
      type: 'okapi',
      path: 'audit-data/:{id}',
      clear: false,
    },
  });

  getInstance() {
    const { resources, match: { params: { id } } } = this.props;
    const selectedInstance = (resources.selectedInstance || {}).records || [];
    if (!selectedInstance || selectedInstance.length === 0 || !id) return null;
    return selectedInstance.find(u => u.id === id);
  }

  render() {
    const records = this.getInstance();
    if (!records) {
      return (
        <Pane
          id="pane-recorddetails"
          defaultWidth={this.props.paneWidth}
          paneTitle="Audit Data"
          lastMenu={<span />}
          dismissible
          onClose={this.props.onClose}
        >
          <div style={{ paddingTop: '1rem' }}><Icon icon="spinner-ellipsis" width="100px" /></div>
        </Pane>
      );
    }
    return (
      <Pane
        id="pane-recorddetails"
        defaultWidth={this.props.paneWidth}
        paneTitle="Audit Data"
        lastMenu={<span />}
        dismissible
        onClose={this.props.onClose}
      >
        <div>
          <Highlight language="json">
            {JSON.stringify(records, null, 2)}
          </Highlight>
        </div>
      </Pane>
    );
  }
}

ViewAuditData.propTypes = {
  stripes: PropTypes.shape({
    connect: PropTypes.func.isRequired,
    locale: PropTypes.string.isRequired,
    logger: PropTypes.object.isRequired,
    intl: PropTypes.shape({
      formatMessage: PropTypes.func.isRequired,
    }),
  }).isRequired,
  resources: PropTypes.shape({
    selectedInstance: PropTypes.shape({
      records: PropTypes.arrayOf(PropTypes.object),
    }),
  }).isRequired,
  match: PropTypes.shape({
    path: PropTypes.string.isRequired,
    params: PropTypes.shape({
      id: PropTypes.string,
    }),
  }),
  onClose: PropTypes.func,
  paneWidth: PropTypes.string.isRequired
};

export default ViewAuditData;


src/routes/search.js
import React from 'react';
import { get } from 'lodash';
import PropTypes from 'prop-types';
import { makeQueryFunction, SearchAndSort } from '@folio/stripes/smart-components';
import ViewAuditData from '../components/ViewAuditData';
import packageInfo from '../../package';

const filterConfig = [
  {
    label: 'Method',
    name: 'method',
    cql: 'method',
    values: [
      { name: 'POST', cql: 'POST' },
      { name: 'PUT', cql: 'PUT' },
      { name: 'DELETE', cql: 'DELETE' },
      { name: 'PATCH', cql: 'PATCH' },
    ],
    restrictWhenAllSelected: true,
  },
  {
    label: 'Module Result Status Codes',
    name: 'Status Code',
    cql: 'module_result',
    values: [
      { name: '201', cql: '201' },
      { name: '204', cql: '204' },
      { name: '403', cql: '403' },
      { name: '400', cql: '400' },
      { name: '422', cql: '422' },
      { name: '500', cql: '500' },
    ],
    restrictWhenAllSelected: true,
  }
];

const searchableIndexes = [
  { label: 'Target Id', value: 'target_id', makeQuery: term => `(target_id="${term}*")` },
  { label: 'Target Type', value: 'target_type', makeQuery: term => `(target_type="${term}*")` },
  { label: 'Request Id', value: 'request_id', makeQuery: term => `(request_id="${term}*")` },
  { label: 'User', value: 'user', makeQuery: term => `(user="${term}*")` },
  { label: 'Timestamp', value: 'timestamp', makeQuery: term => `(timestamp="${term}*")` }
];

class Search extends React.Component {
  static defaultProps = {
    browseOnly: false,
    showSingleResult: true,
  }

  static propTypes = {
    resources: PropTypes.shape({
      records: PropTypes.shape({
        hasLoaded: PropTypes.bool.isRequired,
        other: PropTypes.shape({
          totalRecords: PropTypes.number,
          total_records: PropTypes.number,
        })
      }),
      query: PropTypes.shape({
        qindex: PropTypes.string,
        query: PropTypes.string,
      }),
    }),
    match: PropTypes.shape({
      path: PropTypes.string.isRequired,
    }).isRequired,
    mutator: PropTypes.shape({
      query: PropTypes.shape({
        update: PropTypes.func.isRequired,
      }),
    }),
    stripes: PropTypes.shape({
      logger: PropTypes.shape({
        log: PropTypes.func.isRequired,
      }).isRequired,
    }).isRequired,
    showSingleResult: PropTypes.bool, // eslint-disable-line react/no-unused-prop-types
    browseOnly: PropTypes.bool,
  }

  static manifest = Object.freeze({
    resultCount: { initialValue: 30 },
    query: {
      initialValue: {
        query: '',
        filters: '',
        sort: 'timestamp',
      },
    },
    records: {
      type: 'okapi',
      records: 'audit',
      path: 'audit-data',
      recordsRequired: '%{resultCount}',
      perRequest: 30,
      GET: {
        params: {
          query: makeQueryFunction(
            'cql.allRecords=1',
            'target_id="%{query.query}*"',
            {
              'id': 'target_id',
              'Timestamp': 'timestamp',
              'Target Id': 'target_id',
              'Target Type': 'target_type',
              'User': 'user',
            },
            filterConfig,
            false,
          ),
        },
        staticFallback: { params: {} },
      },
    },
  });

  onChangeIndex = (e) => {
    const qindex = e.target.value;
    const logger = this.props.stripes.logger;
    logger.log('action', `changed query-index to '${qindex}'`);
    this.props.mutator.query.update({ qindex });
  }

  render() {
    const { browseOnly } = this.props;
    const resultsFormatter = {
      'User': x => (x.user || []),
      'Target Id': x => (x.target_id || []),
      'Target Type': x => (x.target_type || []),
      'Timestamp': x => (x.timestamp || [])
    };

    return (<SearchAndSort
      packageInfo={packageInfo}
      objectName="audit"
      selectedIndex={get(this.props.resources.query, 'qindex')}
      searchableIndexes={searchableIndexes}
      searchableIndexesPlaceholder={null}
      onChangeIndex={this.onChangeIndex}
      maxSortKeys={1}
      filterConfig={filterConfig}
      initialResultCount={30}
      resultCountIncrement={30}
      viewRecordComponent={ViewAuditData}
      visibleColumns={['Timestamp', 'Target Id', 'Target Type', 'User']}
      columnWidths={{ 'timestamp': '30%', 'target id': '30%', 'target type': '30%', 'user': '30%' }}
      resultsFormatter={resultsFormatter}
      viewRecordPerms="audit.item.get"
      disableRecordCreation
      parentResources={this.props.resources}
      parentMutator={this.props.mutator}
      path={`${this.props.match.path}/(view)/:id`}
      notLoadedMessage="Enter search query to show results"
      browseOnly={browseOnly}
    />);
  }
}

export default Search;
